File: assembler/Statement.rb
require_relative 'Header'

# requires: description, type, args
class DefineStatement < ContentLoader
  attr_accessor :name, :args

  def getInstance(args)
    Statement.new @name, @args, args
  end

  # balls.
  class Statement
    def initialize(name, args, vars)
      @name = name
      @args = args
      @argsnames = []
      @args.each_key do |a|
        @argsnames << a
        @args[a] = vars[a]
        puts "assigned #{vars[a]} to #{a}" if DEBUG_MODE
      end
      puts "args names: #{@argsnames}"
    end

    def ==(other)
      @content == other.content
    end
  end
end

File: Assembler/Statement.py
from Assembler.macros import DEBUG_MODE
from Assembler.utils import Initializer

# the actual statement that will be considered in the proof process
class StatementInstance:
	def __init__(self,vals,args,name):
		self.name = name
		self.args = {}
		for a in args:
			self.args[a] = vals[a]
			if DEBUG_MODE:
				print(f"assigned {vals[a]} to {a}")

	# overload comparison. Statement comparison will only check for matching variable names
	def __eq__(self, other):
		return self.args == other.args

	# in case attribute is not found
	def __getattr__(self, item):
		return None

# statement info container interface
class Statement(Initializer):
	def __init__(self, content, name):
		Initializer.initialize(self, content, name)

	def getInstance(self, args):
		return StatementInstance(args, self.args, self.name)

# statements loader interface (STATIC). Will create a StatementInstance instance with requested name and args
class DeclaredStatements:
	statements = {}

	@staticmethod
	def init(content):
		for name, s in content.items():
			DeclaredStatements.statements[name] = Statement(s, name)
			print(f"new statement {name}")

	@staticmethod
	def getInstance(name, args):
		print(f"looking for statement {name} to instance...")
		return DeclaredStatements.statements.get(name).getInstance(args)
